src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java

```diff
@@ -174,21 +173,221 @@ public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("select distinct team.students from Team team where team.exercise.course.id = :#{#courseId} and team" + ".shortName = :#{#teamShortName}")
     Set<User> findAllInTeam(@Param("courseId") Long courseId, @Param("teamShortName") String teamShortName);
 
+    /**
+     * Creates the specification to match the provided search term.
+     *
+     * @param searchTerm term to match
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getSearchTermSpecification(String searchTerm) {
+        String extendedSearchTerm = "%" + searchTerm + "%";
+        return (root, query, criteriaBuilder) -> {
+            Predicate login = criteriaBuilder.like(root.get(User_.LOGIN), extendedSearchTerm);
+            Predicate email = criteriaBuilder.like(root.get(User_.EMAIL), extendedSearchTerm);
+            Predicate firstName = criteriaBuilder.like(root.get(User_.FIRST_NAME), extendedSearchTerm);
+            Predicate lastName = criteriaBuilder.like(root.get(User_.LAST_NAME), extendedSearchTerm);
+
+            return criteriaBuilder.or(login, email, firstName, lastName);
+        };
+    }
+
+    /**
+     * Creates the specification to match the provides authorities.
+     *
+     * @param authorities set of possible authorities
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAllUsersMatchingAuthorities(Set<String> authorities) {
+        return (root, query, criteriaBuilder) -> {
+            Subquery<Long> subQuery = query.subquery(Long.class);
+            Root<User> subUserRoot = subQuery.from(User.class);
+            Join<User, Authority> joinedAuthorities = subUserRoot.join(User_.AUTHORITIES, JoinType.LEFT);
+
+            subQuery.select(subUserRoot.get(User_.ID))
+                    .where(criteriaBuilder.and(criteriaBuilder.in(joinedAuthorities.get(Authority_.NAME)).value(authorities),
+                            criteriaBuilder.equal(subUserRoot.get(User_.ID), root.get(User_.ID))))
+                    .groupBy(subUserRoot.get(User_.ID)).having(criteriaBuilder.equal(criteriaBuilder.count(joinedAuthorities), authorities.size()));
+
+            return criteriaBuilder.and(criteriaBuilder.exists(subQuery));
+        };
+    }
+
+    /**
+     * Creates the specification to match the empty authorities.
+     *
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAllUsersMatchingEmptyAuthorities() {
+        return (root, query, criteriaBuilder) -> criteriaBuilder.isEmpty(root.get(User_.AUTHORITIES));
+    }
+
+    /**
+     * Returns the matching authority specification.
+     * @param authorities provided authorities
+     * @return specification used to chain database operations
+     */
+    private Specification<User> getAuthoritySpecification(Set<String> authorities) {
+        if (authorities.contains("NO_AUTHORITY")) {
```

Comment:
Can you use some constants instead of strings here and below?

Reply:
Sure :D